src/pages/workspaces/workspace/analysis/file-utils.ts

```diff
@@ -0,0 +1,65 @@
+import _ from 'lodash/fp'
+import { Ajax } from 'src/libs/ajax'
+import * as Utils from 'src/libs/utils'
+import { ToolLabel, tools, toolToExtensionMap } from 'src/pages/workspaces/workspace/analysis/tool-utils'
+import { CloudProviderType } from 'src/pages/workspaces/workspace/workspace-utils'
+
+// This syntax gives us nominal typing
+// Nominal types are not interchangable even if the underlying type is the same
```

Comment:
As @kyuksel mentioned, it might be worth adding a helper type for this to encode the `__typeToken` convention.

Something like `type NominalType<BaseType, Name extends string> = BaseType & { __typeToken: Name }`

Reply:
Yes, and I also wonder if `string` could also be a generic type `T` so the construct can be used for wrapping a value of any type, not just `string`. Ideally, another extension would also allow multiple values to be wrapped, in which case we'd have something fairly similar to [Scala case classes](https://docs.scala-lang.org/tour/case-classes.html), which would be delightful  😊

Reply:
`BaseType` is that generic type in my example.

Reply:
Why is there `string` in it?

Reply:
That's requiring that the `__typeToken`, which is like an arbitrary ID for the type, be a string. So you could do something like:
```ts
type MyNumberType = NominalType<number, 'My Number Type'>
```
but not
```ts
const MyType = NominalType<string, 3>
```

Reply:
Ah so the fact that `__typeToken` is a `string` would be hidden from clients of `NominalType`, correct?

Reply:
hmm.. where do we think this nominal type definition should live? Maybe we leave it here for now, and can move it to a type-commons or the like down the line...

Reply:
If this is something folks see a lot of value in, I would recommend the __typeToken be a Symbol, which enforces uniqueness across the codebase.